src/models/event-timeline-set.ts

```diff
@@ -842,6 +842,31 @@ export class EventTimelineSet extends TypedEventEmitter<EmittedEvents, EventTime
         // the timelines are not contiguous.
         return null;
     }
+
+    /**
+     * Determine whether a given event can sanely be added to this event timeline set,
+     * for timeline sets relating to a thread, only return true for events in the same
+     * thread timeline, for timeline sets not relating to a thread only return true
+     * for events which should be shown in the main room timeline.
+     * Requires the `room` property to have been set at EventTimelineSet construction time.
+     *
+     * @param event {MatrixEvent} the event to check whether it belongs to this timeline set.
+     * @throws {Error} if `room` was not set when constructing this timeline set.
+     * @return {boolean} whether the event belongs to this timeline set.
+     */
+    public eventBelongs(event: MatrixEvent): boolean {
+        if (!this.room) {
+            throw new Error("Cannot call `EventTimelineSet::eventBelongs without a `room` set. " +
+                "Set the room when creating the EventTimelineSet to call this method.");
+        }
+
+        const { threadId, shouldLiveInRoom } = this.room.eventShouldLiveIn(event);
+
+        if (this.thread) {
+            return this.thread.id === threadId;
+        }
+        return shouldLiveInRoom;
+    }
```

Comment:
We should explain these scenarios that we need to worry about in the JSDoc here. The function description tries to describe this already but we're not able to keep all of these in our head and bugs are creeping in.

 1. If main timeline and thread event, return `undefined`
 1. If thread timeline and main timeline event, return `undefined`
 1. If thread timeline and event from different thread, return `undefined`
 1. If thread timeline and `threadRoot` from different thread, return `undefined`
 1. If the main timeline and main timeline event, return timeline
 1. If thread timeline and event from same thread, return timeline
 1. If main timeline and `threadRoot`, return timeline
 1. If thread timeline and `threadRoot` from same thread, return timeline
 
 Tests cross-referencing the numbered list from the JSDoc would also be good.